pkg/controller/rds/rdsinstance.go

```diff
@@ -67,186 +52,150 @@ type Reconciler struct {
-		runtimev1alpha1.ResourceCredentialsSecretEndpointKey: []byte(instance.Status.Endpoint),
-	}); err != nil {
-		return r.fail(instance, err)
+	cr.Status.AtProvider = rds.GenerateObservation(instance)
+
+	switch cr.Status.AtProvider.DBInstanceStatus {
+	case string(v1alpha2.RDSInstanceStateAvailable):
+		cr.Status.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(cr)
+	case string(v1alpha2.RDSInstanceStateCreating):
+		cr.Status.SetConditions(runtimev1alpha1.Creating())
+	case string(v1alpha2.RDSInstanceStateDeleting):
+		cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		cr.Status.SetConditions(runtimev1alpha1.Unavailable())
 	}
 
-	instance.Status.SetConditions(runtimev1alpha1.ReconcileSuccess())
-	return result, r.Update(ctx, instance)
+	return resource.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  rds.IsUpToDate(cr.Spec.ForProvider, instance),
+		ConnectionDetails: rds.GetConnectionDetails(*cr),
+	}, nil
 }
 
-func (r *Reconciler) _delete(instance *databasev1alpha2.RDSInstance, client rds.Client) (reconcile.Result, error) {
-	instance.Status.SetConditions(runtimev1alpha1.Deleting())
-
-	if instance.Spec.ReclaimPolicy == runtimev1alpha1.ReclaimDelete {
-		if _, err := client.DeleteInstance(instance.Status.InstanceName); err != nil && !rds.IsErrorNotFound(err) {
-			return r.fail(instance, err)
-		}
+func (e *external) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha2.RDSInstance)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errNotRDSInstance)
 	}
-
-	meta.RemoveFinalizer(instance, finalizer)
-	instance.Status.SetConditions(runtimev1alpha1.ReconcileSuccess())
-	return result, r.Update(ctx, instance)
-}
-
-// Reconcile reads that state of the cluster for a Instance object and makes changes based on the state read
-// and what is in the Instance.Spec
-func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	log.V(logging.Debug).Info("reconciling", "kind", databasev1alpha2.RDSInstanceKindAPIVersion, "request", request)
-	// Fetch the CRD instance
-	instance := &databasev1alpha2.RDSInstance{}
-
-	err := r.Get(ctx, request.NamespacedName, instance)
+	// generate new password
+	password, err := util.GeneratePassword(20)
 	if err != nil {
-		if kerrors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
-			return reconcile.Result{}, nil
-		}
-		log.Error(err, "failed to get object at start of reconcile loop")
-		return reconcile.Result{}, err
+		return resource.ExternalCreation{}, err
 	}
-
-	rdsClient, err := r.connect(instance)
-	if err != nil {
-		return r.fail(instance, err)
+	req := e.client.CreateDBInstanceRequest(rds.GenerateCreateDBInstanceInput(meta.GetExternalName(cr), password, &cr.Spec.ForProvider))
+	req.SetContext(ctx)
+	_, err = req.Send()
```

Comment:
How does this behave if the RDS instance already exists? Are we intending to fail the reconcile in this case?

Taking a quick skim of the v1beta1 controllers, I think that RDS, ReplicationGroup, and CloudMemorystoreInstance would return an error if trying to create an instance that already exists (which I recall is now our desired behaviour), but it looks like CloudSQL would not. Should I open a PR to update CloudSQL?

Reply:
https://github.com/crossplaneio/stack-gcp/pull/62

Reply:
Yes, thanks for the PR. It's a little bit disturbing to have that behavior difference by I think it will be addressed with external policies once and for all.